### PR TITLE
update serial number to a valid non-zero number in ca certificate

### LINKED ISSUE #117790: all cluster CA certificates created when a cluster is created have an empty Serial number
### What happened?

Hi I am from Amazon EKS. our customer recently sent a query to us that they found all cluster CA certificates created when an EKS cluster is created have an empty Serial number. 

After investigation, I found the serial number is actually set to 0 by https://github.com/kubernetes/client-go/blob/master/util/cert/cert.go#L61 through kubeadm init phase certs.

Shall I tell customer that this is by design intentionally, or is this some issue customer found and need to be updated?

thank you

### What did you expect to happen?

we want to confirm if we shall tell customer that serial number is set to 0 by design intentionally or it is some issue.

### How can we reproduce it (as minimally and precisely as possible)?

kubeadm init phase certs all

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
1.24
```

</details>

### Cloud provider

<details>
aws
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by nnmin-aws**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
EKS customer complains all cluster CA certificates created when an EKS cluster is created have an empty Serial number.

#### Which issue(s) this PR fixes:
fixes https://github.com/kubernetes/kubernetes/issues/117790

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 1]
```diff
@@ -57,8 +58,12 @@ type AltNames struct {
 // NewSelfSignedCACert creates a CA certificate
 func NewSelfSignedCACert(cfg Config, key crypto.Signer) (*x509.Certificate, error) {
 	now := time.Now()
+	serial, err := cryptorand.Int(cryptorand.Reader, new(big.Int).SetInt64(math.MaxInt64))
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/cert/cert.go:61]: actually, can't this still return zero?

> Int returns a uniform random value in [0, max). It panics if max <= 0.



/hold

**Comment by neolit123** [comment]
please change this in the PR description

> #117790

to

> fixes #117790

so that the related issue is closed.

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 1 above)
Comment (dims) [reply staging/src/k8s.io/client-go/util/cert/cert.go:61]: Ah fun!

@liggitt this snippet was copied from elsewhere in our repo :)

https://cs.k8s.io/?q=serial%2C%20err%20%3A%3D.*rand&i=nope&files=&excludeFiles=&repos=kubernetes/kubernetes

Comment (neolit123) [reply staging/src/k8s.io/client-go/util/cert/cert.go:61]: we could fix the existing problem as part of this PR in a separate commit?

Comment (nnmin-aws) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: Could someone help take a look at the update? if it is good, I will fix the existing problem in a separate commit with similar code.

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 2]
```diff
@@ -656,9 +657,10 @@ func NewSignedCert(cfg *CertConfig, key crypto.Signer, caCert *x509.Certificate,
 			CommonName:   cfg.CommonName,
 			Organization: cfg.Organization,
 		},
-		DNSNames:              cfg.AltNames.DNSNames,
-		IPAddresses:           cfg.AltNames.IPs,
-		SerialNumber:          serial,
+		DNSNames:    cfg.AltNames.DNSNames,
+		IPAddresses: cfg.AltNames.IPs,
+		// add 1 to serial to make it a uniform random value in [1, max).
+		SerialNumber:          new(big.Int).Add(serial, big.NewInt(1)),
```
Comment (neolit123) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:663]: my suggestion would be to prepare `serial` with the  +1 in advance and just have the `SerialNumber: serial` in cert structs. also write a single comment explaining briefly the range `[1, max-1]` generation.

applies to other locations too.

**Review by neolit123 — COMMENTED** [review]
/lgtm

**Comment by dims** [comment]
/release-note-none
/hold cancel

@liggitt 's concern in https://github.com/kubernetes/kubernetes/pull/117791#discussion_r1185627212 has been fixed. (for all locations the same kind of snippet is used!)

**Comment by liggitt** [comment]
lgtm

**Comment by champtar** [comment]
Created the cherry pick PRs for this PR as I get a git conflict while trying to backport #118922

**Pull request merged**